pkg/operator/configobservation/auth/external_oidc.go

```diff
@@ -0,0 +1,159 @@
+}
+
+// ObserveExternalOIDC observes the authentication.config/cluster resource
+// and if the type field is set to OIDC, it configures an external OIDC provider
+// to the KAS pods by setting the --authentication-config apiserver argument. It also
+// takes care of synchronizing the structured auth config file into the apiserver's namespace
+// so that it gets mounted as a static file on each node.
+func (o *externalOIDC) ObserveExternalOIDC(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	if o.featureGateAccessor == nil || !o.featureGateAccessor.AreInitialFeatureGatesObserved() {
+		// if we haven't observed featuregates yet, return the existing
+		return existingConfig, nil
+	}
+
+	featureGates, err := o.featureGateAccessor.CurrentFeatureGates()
+	if err != nil {
+		return existingConfig, []error{err}
+	}
+
+	if !featureGates.Enabled(features.FeatureGateExternalOIDC) {
+		return existingConfig, nil
+	}
+
+	listers := genericListers.(configobservation.Listers)
+	auth, err := listers.AuthConfigLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		klog.Warningf("authentications.config.openshift.io/cluster: not found")
+		return existingConfig, nil
+	} else if err != nil {
+		return existingConfig, []error{err}
+	}
+
+	if auth.Spec.Type != configv1.AuthenticationTypeOIDC {
+		if _, err := listers.ConfigMapLister().ConfigMaps(operatorclient.TargetNamespace).Get(AuthConfigCMName); errors.IsNotFound(err) {
+			return nil, nil
+
+		} else if err != nil {
+			return existingConfig, []error{fmt.Errorf("failed to get configmap %s/%s: %v", operatorclient.TargetNamespace, AuthConfigCMName, err)}
+		}
+
+		// empty source name/namespace effectively deletes target configmap
+		if err := syncConfigMap(genericListers.ResourceSyncer(), "", "", recorder); err != nil {
+			return existingConfig, []error{err}
+		}
+
+		return nil, nil
+	}
+
+	// auth type is OIDC
+
+	sourceAuthConfig, err := listers.ConfigMapLister().ConfigMaps(SourceAuthConfigCMNamespace).Get(AuthConfigCMName)
+	if errors.IsNotFound(err) {
+		klog.Warningf("configmap %s/%s not found; skipping configuration of OIDC", SourceAuthConfigCMNamespace, AuthConfigCMName)
+		return existingConfig, nil
+
+	} else if err != nil {
+		return existingConfig, []error{fmt.Errorf("failed to get configmap %s/%s: %v", SourceAuthConfigCMNamespace, AuthConfigCMName, err)}
+	}
+
+	if err := validateAuthConfigCM(sourceAuthConfig); err != nil {
+		return existingConfig, []error{fmt.Errorf("configmap %s/%s is invalid: %v", SourceAuthConfigCMNamespace, AuthConfigCMName, err)}
+	}
+
+	targetAuthConfig, err := listers.ConfigMapLister().ConfigMaps(operatorclient.TargetNamespace).Get(AuthConfigCMName)
+	if err != nil && !errors.IsNotFound(err) {
+		return existingConfig, []error{err}
+	}
+
+	if targetAuthConfig == nil || !equality.Semantic.DeepEqual(targetAuthConfig.Data, sourceAuthConfig.Data) {
```

Comment:
I would expect the syncer to be responsible for deciding whether or not it needs to do a write. Why test here?

Reply:
As per my previous comment, I wanted to avoid an extra API call that the syncer does.